src/Sarif/Visitors/SarifCurrentToVersionOneVisitor.cs

```diff
@@ -1032,6 +1014,81 @@ private static string CreateFileDictionaryKey(FileData v2File, IList<FileData> v
             return logicalLocationsVersionOne;
         }
 
+        // In SARIF v1, run.resources.rules was a dictionary, whereas in v2 it is an array.
+        // Normally, the lookup key into the v1 rules dictionary is the rule id. But some
+        // tools allow multiple rules to have the same id. In that case we must synthesize
+        // a unique key for each rule with that id. We choose "<ruleId>-<n>", where <n> is
+        // 1 for the second occurrence, 2 for the third, and so on.
+        private static IDictionary<int, string> CreateV2RuleIndexToV1KeyMapping(IList<Rule> rules)
+        {
+            var v2RuleIndexToV1KeyMap = new Dictionary<int, string>();
+
+            if (rules != null)
+            {
+                // Keep track of how many distinct rules have each id.
+                var ruleIdToCountMap = new Dictionary<string, int>();
+
+                for (int i = 0; i < rules.Count; ++i)
+                {
+                    Rule rule = rules[i];
+                    if (rule.Id != null)
+                    {
+                        ruleIdToCountMap[rule.Id] = ruleIdToCountMap.ContainsKey(rule.Id)
+                            ? ruleIdToCountMap[rule.Id] + 1
+                            : 1;
+
+                        v2RuleIndexToV1KeyMap[i] = ruleIdToCountMap[rule.Id] == 1
+                            ? null
+                            : rule.Id + '-' + (ruleIdToCountMap[rule.Id] - 1).ToString();
+                    }
+                }
+            }
+
+            return v2RuleIndexToV1KeyMap;
+        }
+
+        private static string GetV1RuleKeyFromV2Index(
+            int ruleIndex,
+            IDictionary<int, string> v2RuleIndexToV1KeyMap)
+        {
+            string ruleKey = null;
+            if (0 <= ruleIndex && ruleIndex < v2RuleIndexToV1KeyMap?.Count)
```

Comment:
>0 <= ruleIndex [](start = 16, length = 14)

can you just make this ruleIndex > -1 perhaps? your comparison here is confusing. if you retain it, please comment the code.

Reply:
I've got a principled objection to treating a sentinel value numerically just because it happens to be ordered properly. `ruleIndex != -1` ("it's not the sentinel!") is tempting, but it lets `-2` slip through. I like `ruleIndex >= 0` because it states the validity condition independent of what we happened to choose as the sentinel.

Or were you objecting to the idiom `min <= x && x < limit` (which I occasionally use to mimic the math notation `min < x < limit`)?

---
In reply to: [249574859](https://github.com/Microsoft/sarif-sdk/pull/1228#discussion_r249574859) [](ancestors = 249574859)

Reply:
Marking this one "By Design" (because I really do like the comparison the way I wrote it), but marking the next one "Resolved" because you're right, we don't need the comparison at all.

---
In reply to: [249613504](https://github.com/Microsoft/sarif-sdk/pull/1228#discussion_r249613504) [](ancestors = 249613504,249574859)